src/L2OALM.jl

```diff
@@ -1,5 +1,426 @@
+- `hyperparameters`: Dictionary of hyperparameters and hyper-states for the training step.
+- `parameter_update_fns`: Vector of functions to update hyperparameters after each training step.
+- `reconcile_state`: Function to reconcile the state after processing a batch of data.
+"""
+mutable struct TrainingStepLoop
+    loss_fn::Function
+    stopping_criteria::Vector{Function}
+    hyperparameters::Dict{Symbol,Any}
+    parameter_update_fns::Vector{Function}
+    reconcile_state::Function
+    pre_hook::Function
+end
+
+"""
+    _pre_hook_primal(θ, primal_model, train_state_primal, dual_model, train_state_dual, bm)
+
+Default pre-hook function for the primal model in the L2O-ALM algorithm.
+This function performs a forward pass through the dual model to obtain the dual predictions.
+"""
+function _pre_hook_primal(
+    θ,
+    primal_model,
+    train_state_primal,
+    dual_model,
+    train_state_dual,
+    bm,
+)
+    # Forward pass for dual model
+    dual_hat_k, _ = dual_model(θ, train_state_dual.parameters, train_state_dual.states)
+
+    return (dual_hat_k,)
+end
+
+"""
+    _pre_hook_dual(θ, primal_model, train_state_primal, dual_model, train_state_dual, bm)
+
+Default pre-hook function for the dual model in the L2O-ALM algorithm.
+This function performs a forward pass through the primal model to obtain the predicted state and constraints.
+"""
+function _pre_hook_dual(
+    θ,
+    primal_model,
+    train_state_primal,
+    dual_model,
+    train_state_dual,
+    bm,
+)
+    # # Forward pass for primal model
+    X̂, _ = primal_model(θ, train_state_primal.parameters, train_state_primal.states)
+    gh = constraints!(bm, X̂, Θ)
+
+    # Forward pass for dual model
+    dual_hat, _ = dual_model(θ, train_state_dual.parameters, train_state_dual.states)
+
+    return (dual_hat, gh)
+end
+
+"""
+    _reconcile_alm_primal_state(batch_states::Vector{NamedTuple})
+
+Default function that reconciles the state of the primal model after processing a batch of data.
+This function computes the maximum violation, mean violations, mean objectives, and total loss
+from the batch states.
+"""
+function _reconcile_alm_primal_state(batch_states::Vector{NamedTuple})
+    max_violation = maximum([s.new_max_violation for s in batch_states])
+    mean_violations = mean([s.mean_violations for s in batch_states])
+    mean_objs = mean([s.mean_objs for s in batch_states])
+    mean_loss = mean([s.total_loss for s in batch_states])
+    return (;
+        new_max_violation = max_violation,
+        mean_violations = mean_violations,
+        mean_objs = mean_objs,
+        total_loss = mean_loss,
+    )
+end
+
+"""
+    _reconcile_alm_dual_state(batch_states::Vector{NamedTuple})
+
+Default function that reconciles the state of the dual model after processing a batch of data.
+This function computes the mean dual loss from the batch states.
+"""
+function _reconcile_alm_dual_state(batch_states::Vector{NamedTuple})
```

Comment:
```suggestion
function _reconcile_dual_state(batch_states::Vector{NamedTuple})
```

`alm` can be removed since that is this whole repo :smile: (needs updates everywhere else, and for the primal version, update_rho, etc. too. let me know if you agree and I can add that commit)